### PR TITLE
Feature/pps sync

### PR DESCRIPTION
**Pull request by rikba**
Timestamping sensor messages on arrival introduces jitter (and transmission and buffering delay) from the communication interface (USB, TCP, Serial). In this PR we address this problem by timestamping the GPS data with the correct GPS timestamp and synchronizing the (x86) host against the PPS clock.

# Contributions
- A kernel module that is configurable through modprobe to readout the PPS time on an x86 host
- Chrony settings to synchronize against the PPS and NMEA
- GPS time stamping in the Piksi ROS driver (configurable through ROS param)
- IMU and MAG scaling and publishing
- Evaluation script
- Installation script for all of the above
- Rover settings

# Configuration
To activate GPS time stamping set the [flag in the driver settings](https://github.com/ethz-asl/ethz_piksi_ros/blob/feature/pps_sync/piksi_multi_rtk_ros/cfg/piksi_multi_driver_settings.yaml#L71).

# Hardware Setup
The Piksi is delivering its SBP messages through TCP to the ROS driver. UART 1 publishes the NMEA GPZDA message at 1Hz to get absolute time on the computer. The PPS is enabled with default settings and goes into a GPIO port of the x86.   
For validation we record an additional IMU on a microcontroller connected to the host computer where the hardware timestamps get translated via [cuckoo time translator](https://github.com/ethz-asl/cuckoo_time_translator).

## System Diagram
pps_piksi

## Piksi UART 1 Settings
Screenshot_2019-05-20_18-45-30

## Piksi NMEA Settings
Screenshot_2019-05-20_18-45-40

# Software Setup
On the host machine we install a [custom PPS driver](https://github.com/ethz-asl/ethz_piksi_ros/blob/feature/pps_sync/piksi_pps_sync/install/install.sh#L7-L16) to expose the PPS signal to Chrony. The NMEA is read out through [GPSD](https://github.com/ethz-asl/ethz_piksi_ros/blob/feature/pps_sync/piksi_pps_sync/install/install.sh#L18-L30). Chrony is configured to use the [PPS as a local reference clock](https://github.com/ethz-asl/ethz_piksi_ros/blob/feature/pps_sync/piksi_pps_sync/install/install.sh#L32-L38).

**NOTE:** On an ARM host computer you can easily [configure the default PPS-GPIO module](http://www.catb.org/gpsd/gpsd-time-service-howto.html) to readout the PPS signal.  

# Results
We validate synchronization of the host against the GPS clock using `chronyd sources -v`. After some convergence time one can observe nanoseconds offset between the PPS and the host clock.

To show the improvement over the existing time stamp on arrival (`rospy.Time.now()`) we cross validate the angular velocity of the Piksi IMU with the external IMU described above. Here the system was rotated on a chair.

## Not synchronizing against PPS
2019-05-22-09-47-12_no_sync
## Clear buffering delays in the time stamps
2019-05-22-09-47-12 bag_no_sync_buffering

## Synchronizing against PPS
2019-05-22-11-54-47_sync
## No more buffering delays in the time stamps
2019-05-22-11-54-47_sync_no_buffering

First of all we observe that the synchronized messages still have a delay of 10-15 ms. This may be due to internal filtering of the Piksi Bosch IMU, transmission delays or translation delay of the external ADIS IMU. But very importantly there is **no buffering delay / jitter on the data anymore**!

This PR may also be relevant to @floriantschopp @hitimo @inkyusa @fmina @HannesSommer @FMTHUBurr @Schartel @danielhentzen @nrjl @mfehr

### CODE REVIEW

File: piksi_multi_rtk_ros/src/piksi_multi_rtk_ros/piksi_multi.py [hunk 1]
```diff
@@ -161,6 +163,13 @@ def __init__(self):
 
         # Other parameters.
         self.publish_raw_imu_and_mag = rospy.get_param('~publish_raw_imu_and_mag', False)
+        # Publish IMU
+        self.acc_scale = 8 * 9.81 / 32768
```
Comment (marco-tranzatto) [inline piksi_multi_rtk_ros/src/piksi_multi_rtk_ros/piksi_multi.py]: is `32768` always stable or would you rather use a constant here?

File: piksi_pps_sync/package.xml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0"?>
+<package format="2">
+  <name>piksi_pps_sync</name>
+  <version>0.0.0</version>
+  <description>Tools to synchronize against the GPS clock.</description>
+
+  <author email="brik@ethz.ch">Rik Bähnemann</author>
+  <maintainer email="brik@ethz.ch">Rik Bähnemann</maintainer>
+  <license>TBD</license>
```
Comment (marco-tranzatto) [inline piksi_pps_sync/package.xml]: same license as whole repo?

File: piksi_pps_sync/package.xml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+  <buildtool_depend>catkin</buildtool_depend>
+  <buildtool_depend>catkin_simple</buildtool_depend>
```
Comment (marco-tranzatto) [inline piksi_pps_sync/package.xml]: other packages in this repo do not use catkin_simple. Try to pull in another dependency if possible.

File: piksi_multi_rtk_ros/cfg/piksi_multi_driver_settings.yaml [hunk 4]
```diff
@@ -64,3 +64,8 @@ tcp_addr: 192.168.0.222
 # tcp_port: the port number of the TCP server. By default, Swift Navigation
 # receivers publish SBP data on TCP port 55555.
 tcp_port: 55555
+
+# Time settings.
+# use_gps_time: If true, it timestamps the position information with GPS time.
+# Otherwise uses rospy.Time.now()
+use_gps_time: true
```
Comment (marco-tranzatto) [inline piksi_multi_rtk_ros/cfg/piksi_multi_driver_settings.yaml]: shouldn't this be false per default, as standard users may not want to use this new feature?

File: piksi_multi_rtk_ros/src/piksi_multi_rtk_ros/piksi_multi.py (hunk 1 above)
Comment (rikba) [reply piksi_multi_rtk_ros/src/piksi_multi_rtk_ros/piksi_multi.py]: Made the fixed scaling constants constants.

File: piksi_pps_sync/package.xml (hunk 3 above)
Comment (rikba) [reply piksi_pps_sync/package.xml]: Done.

File: piksi_multi_rtk_ros/cfg/piksi_multi_driver_settings.yaml (hunk 4 above)
Comment (rikba) [reply piksi_multi_rtk_ros/cfg/piksi_multi_driver_settings.yaml]: Changed default to false.

File: piksi_pps_sync/package.xml (hunk 2 above)
Comment (rikba) [reply piksi_pps_sync/package.xml]: Done.

**Pull request merged**